cpp/src_prims/distance/fused_l2_nn.h

```diff
@@ -0,0 +1,591 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_utils.h>
+#include <stdint.h>
+#include <cub/cub.cuh>
+#include <limits>
+
+namespace MLCommon {
+namespace Distance {
```

Comment:
This was a tricky one for me! This prim fuses L2-distance (which belongs to Distance namespace) and 1-NN (which belongs to Selection namespace). Since typically distance computation takes more time than 1-NN, I just kept it under Distance namespace itself.

Reply:
What say?